fmt/mod.c

```diff
@@ -396,3 +468,211 @@ int fmt_mod15_load_song(song_t *song, slurp_t *fp, unsigned int lflags)
 {
 	return fmt_mod_load_song(song, fp, lflags, 1);
 }
+
+/* incomplete 31 sample M.K. amiga mod saving routine */
+int fmt_mod_save_song(disko_t *fp, song_t *song)
+{
+	uint8_t mod_songtitle[20];
+	uint8_t mod_sampleheader[30];
+	uint8_t mod_orders[128];
+	uint8_t tmp[128];
+	uint8_t mod_pattern[1024];
+
+	int nord, nsmp, maxpat, jmax, joutpos;
+	long tmppos;
+	int i, j, n, period;
+	unsigned int warn = 0;
+	song_note_t *m;
+
+	if (song->flags & SONG_INSTRUMENTMODE)
+		warn |= 1 << WARN_INSTRUMENTS;
+	if (song->flags & SONG_LINEARSLIDES)
+		warn |= 1 << WARN_LINEARSLIDES;
+
+	nsmp = csf_get_num_samples(song); // Getting number of samples
+	if (nsmp > 31) {
+		nsmp = 31;
+		warn |= 1 << WARN_MAXSAMPLES;
+	}
+
+	if (3 < csf_get_highest_used_channel(song))
+		warn |= 1 << WARN_MAXCHANNELS;
+
+	memcpy(mod_songtitle, song->title, 20);
+	disko_write(fp, mod_songtitle, 20); // writing song title
+
+	// Now writing sample headers
+	for(n = 1; n <= 31; ++n) {
+		for(i = 0; i < 30; ++i) mod_sampleheader[i] = i / 29;
+		if(n <= nsmp) {
+			if(song->samples[n].global_volume != 64)
+				warn |= 1 << WARN_SAMPLEVOL;
+			if((song->samples[n].flags & (CHN_LOOP | CHN_PINGPONGLOOP)) == (CHN_LOOP | CHN_PINGPONGLOOP) || (song->samples[n].flags & CHN_SUSTAINLOOP))
+				warn |= 1 << WARN_LOOPS;
+			if(song->samples[n].vib_depth != 0)
+				warn |= 1 << WARN_SAMPLEVIB;
+			memcpy(mod_sampleheader, song->samples[n].name, 22); // sample name
+			if(song->samples[n].length <= 0x1FFFE) {
+				mod_sampleheader[22] = song->samples[n].length >> 9; // sample 11th word MSB length/2
+				mod_sampleheader[23] = song->samples[n].length >> 1; // sample 11th word LSB length/2
+				if(1 & song->samples[n].length)
+					warn |= 1 << WARN_LONGSAMPLES;
+			} else {
+				mod_sampleheader[22] = 0xFF;
+				mod_sampleheader[23] = 0xFF;
+				warn |= 1 << WARN_LONGSAMPLES;
+			}
+			for(j = 15; j && (finetune_table[j] > song->samples[n].c5speed); --j)
+				if(((song->samples[n].c5speed) > 10000) && (j == 8))
+					break; // determine from finetune_table entry
+			mod_sampleheader[24] = (j ^ 8) & 0x0f; // sample 24th byte finetune value
+			mod_sampleheader[25] = (song->samples[n].volume + 1) / 4; // sample 25th byte sample volume value 0..64 scaled
+			if( (song->samples[n].flags & CHN_LOOP) && (song->samples[n].loop_start < song->samples[n].loop_end) && (song->samples[n].loop_end <= MIN(song->samples[n].length, 0x1FFFE)) ) {
+				mod_sampleheader[26] = song->samples[n].loop_start >> 9;// loop start MSB /2
+				mod_sampleheader[27] = song->samples[n].loop_start >> 1;// loop start LSB /2
+				mod_sampleheader[28] = (song->samples[n].loop_end - song->samples[n].loop_start) >> 9;// loop length MSB /2
+				mod_sampleheader[29] = (song->samples[n].loop_end - song->samples[n].loop_start) >> 1;// loop length LSB /2
+			}
```

Comment:
the `else` block seems to have gone missing here (`mod_sampleheader[29] = 1;` in particular is missing)

Reply:
There was a small refactoring commit, which initialized the *mod_sampleheader with (((unsigned int)(floor(abs(index)/29.0)))) around line 506.

Reply:
Can you please avoid this kind of code golf? It may look cool to you but in the end makes maintenance of the code much more complicated for no good reason.

Reply:
done.

Reply:
Thanks!